lib/ansible/parsing/vault/__init__.py

```diff
@@ -1356,11 +1354,51 @@ def _encrypt_pycrypto(b_plaintext, b_key1, b_key2, b_iv):
 
         return to_bytes(hmac.hexdigest(), errors='surrogate_or_strict'), hexlify(b_ciphertext)
 
+    @staticmethod
+    def _gen_salt(b_plaintext=None, secret=None):
+        # If no secret is provided, we should not generate the salt based on the hash.
+        # If we would do that, we leak the sha256 hash of the plaintext.
+        # This is most of the time not a problem, except if the plaintext contains guessable
+        # content or is partially known by the attacker. The attacker then could simply
+        # broothforce the plaintext content. This is not a problem for larger files, where
+        # the structure is not known by the attacker.
+        #
+        # If we use inline encryption and only encrypt the password of the following variable
+        #   password = '1234'
+        # an attacker could take the sha256 hash out of the salt value
+        # and use it to broothforce the password, as sha256('1234') == salt.
+        #
+        # If a secret was provided, we first append it the the plaintext and than generate the
+        # hash. By doing so we increase the length of the plaintext and also add a string that is
+        # in all cases unknown to the attacker (the passphrase).
+        # So the plaintext hash is now completely useless to the attacker.
+        #
+        # Therefore we also use a randomn salt if no plaintext is provided, or we would leak the
+        # sha256 hash of the secret.
+
+        if not secret or not b_plaintext:
+            return os.urandom(32)
```

Comment:
Since we're falling back to os.urandom in the case of empty, perhaps we should also fall back in the case of not enough entropy in the plaintext?  Is something like len(b_plaintext ) < 256 sufficient for that?

Should a similar check be applied to the secret?

Reply:
<nod>  I see what you mean about checking the length for the secret leaking information about the secret that the attacker would not otherwise know.  However, checking the length for the plaintext doesn't seem to suffer from that same problem.  The attacker already has an idea of how long the plaintext is because the length of the ciphertext will reflect that.  So checking the length of the plaintext and falling back to os.urandom() if the plaintext is too short seems like it can prevent some problems without revealing new information?  (Hmm... it would reveal that the plaintext is less than a precise value whereas looking at the length of the ciphertext  would be more vague.... Would that be important or not?)

The attacker can learn a lot about what is in a vault file from external sources... for instance, since they're yaml files, they likely have a rough outline of:
```
---
var: value
```

an attacker might also find the precise name to use for "var" by looking at playbooks to see which variable names exist that are not defined outside of a playbook.  So then the only thing about the plaintext that they'd be lacking would be the actual value....